velox/expression/tests/ExpressionRunner.cpp

```diff
@@ -129,47 +207,76 @@ void ExpressionRunner::run(
-  } else {
-    inputVector = std::dynamic_pointer_cast<RowVector>(
-        restoreVectorFromFile(inputPath.c_str(), deserializerPool.get()));
-    VELOX_CHECK_NOT_NULL(
-        inputVector,
-        "Input vector is not a RowVector: {}",
-        inputVector->toString());
-    VELOX_CHECK_GT(inputVector->size(), 0, "Input vector must not be empty.");
-  }
-
   fuzzer::InputRowMetadata inputRowMetadata;
   if (!inputRowMetadataPath.empty()) {
     inputRowMetadata = fuzzer::InputRowMetadata::restoreFromFile(
         inputRowMetadataPath.c_str(), pool.get());
   }
 
+  std::vector<fuzzer::InputTestCase> inputTestCases;
+  if (inputPaths.empty()) {
+    inputTestCases.push_back(
+        {std::make_shared<RowVector>(
+             deserializerPool.get(),
+             ROW({}),
+             nullptr,
+             1,
+             std::vector<VectorPtr>{}),
+         SelectivityVector(1)});
+  } else {
+    std::vector<std::string> inputPathsList = split(inputPaths);
+    std::vector<std::string> inputSelectivityPaths =
+        split(inputSelectivityVectorPaths);
+    for (int i = 0; i < inputPathsList.size(); i++) {
+      auto inputVector =
+          std::dynamic_pointer_cast<RowVector>(restoreVectorFromFile(
+              inputPathsList[i].c_str(), deserializerPool.get()));
+      VELOX_CHECK_NOT_NULL(
+          inputVector,
+          "Input vector is not a RowVector: {}",
+          inputVector->toString());
+      VELOX_CHECK_GT(inputVector->size(), 0, "Input vector must not be empty.");
+      if (inputSelectivityPaths.size() > i) {
+        inputTestCases.push_back(
+            {inputVector,
+             restoreSelectivityVectorFromFile(
+                 inputSelectivityPaths[i].c_str())});
+      } else {
+        inputTestCases.push_back(
+            {inputVector, SelectivityVector(inputVector->size(), true)});
+      }
+    }
+    applyModificationsToInput(inputTestCases, inputRowMetadata);
+  }
+
+  VELOX_CHECK(inputTestCases.size() > 0);
+  auto inputRowType = inputTestCases[0].inputVector->type();
+
   parse::registerTypeResolver();
 
   if (mode == "query") {
     core::DuckDbQueryPlanner planner{pool.get()};
+    for (int i = 0; i < inputTestCases.size(); ++i) {
+      auto& testCase = inputTestCases[i];
+      if (inputRowType->size()) {
+        LOG(INFO) << "Registering input vector as table t: "
+                  << inputRowType->toString();
+        planner.registerTable("t", {testCase.inputVector});
+      }
 
-    if (inputVector->type()->size()) {
-      LOG(INFO) << "Registering input vector as table t: "
-                << inputVector->type()->toString();
-      planner.registerTable("t", {inputVector});
-    }
-
-    auto plan = planner.plan(sql);
-    auto results = exec::test::AssertQueryBuilder(plan).copyResults(pool.get());
+      auto plan = planner.plan(sql);
+      auto results =
+          exec::test::AssertQueryBuilder(plan).copyResults(pool.get());
```

Comment:
Are we running the input batches individually, such that we couldn't reproduce bugs that appear from dictionary memoization?

Reply:
good catch, for some reason I thought this was running using duckDB.
One caveat here is that there is no way to plug in the Selectivity vector into the plan without non-trivially changing the input which can alter the test conditions. Let me call out this limitation in the flag's description.